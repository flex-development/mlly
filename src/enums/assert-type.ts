/**
 * @file Enums - AssertType
 * @module mlly/enums/AssertType
 */

/**
 * `import` assertion types.
 *
 * ::: info
 * It's unclear whether or not the HTML spec will require assertion types for
 * WebAssembly modules.
 * :::
 *
 * @see https://github.com/WebAssembly/esm-integration/issues/42
 *
 * @enum {Lowercase<string>}
 */
enum AssertType {
  IMPLICIT = 'javascript',
  JSON = 'json'
}

export default AssertType
