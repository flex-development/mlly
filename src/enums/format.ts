/**
 * @file Enums - Format
 * @module mlly/enums/Format
 */

/**
 * Module formats.
 *
 * @enum {Lowercase<string>}
 */
enum Format {
  BUILTIN = 'builtin',
  COMMONJS = 'commonjs',
  JSON = 'json',
  MODULE = 'module',
  WASM = 'wasm'
}

export default Format
