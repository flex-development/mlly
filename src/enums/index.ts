/**
 * @file Entry Point - Enums
 * @module mlly/enums
 */

export { default as Format } from './format'
export { default as SpecifierKind } from './kind-specifier'
export { default as StatementKind } from './kind-statement'
export { default as SyntaxKind } from './kind-syntax'
