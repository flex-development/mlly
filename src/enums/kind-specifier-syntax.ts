/**
 * @file Enums - SpecifierSyntaxKind
 * @module mlly/enums/SpecifierSyntaxKind
 */

/**
 * Module specifier syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierSyntaxKind {
  DYNAMIC = 'dynamic',
  STATIC = 'static'
}

export default SpecifierSyntaxKind
