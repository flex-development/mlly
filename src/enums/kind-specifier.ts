/**
 * @file Enums - SpecifierKind
 * @module mlly/enums/SpecifierKind
 */

/**
 * Module specifier types.
 *
 * @see https://nodejs.org/api/esm.html#terminology
 *
 * @enum {Lowercase<string>}
 */
enum SpecifierKind {
  ABSOLUTE = 'absolute',
  BARE = 'bare',
  RELATIVE = 'relative'
}

export default SpecifierKind
