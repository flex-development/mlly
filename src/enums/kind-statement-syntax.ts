/**
 * @file Enums - StatementSyntaxKind
 * @module mlly/enums/StatementSyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementSyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default StatementSyntaxKind
