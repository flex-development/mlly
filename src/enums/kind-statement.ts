/**
 * @file Enums - StatementKind
 * @module mlly/enums/StatementKind
 */

/**
 * CommonJS and ESM statement kinds.
 *
 * @enum {Lowercase<string>}
 */
enum StatementKind {
  EXPORT = 'export',
  IMPORT = 'import',
  REQUIRE = 'require'
}

export default StatementKind
