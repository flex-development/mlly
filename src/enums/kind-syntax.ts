/**
 * @file Enums - SyntaxKind
 * @module mlly/enums/SyntaxKind
 */

/**
 * CommonJS and ESM statement syntax kinds.
 *
 * @enum {Lowercase<string>}
 */
enum SyntaxKind {
  DECLARATION = 'declaration',
  DEFAULT = 'default',
  DEFAULT_WITH_NAMED = 'default-with-named',
  DEFAULT_WITH_NAMESPACE = 'default-with-namespace',
  DYNAMIC = 'dynamic',
  LIST = 'list',
  NAMED = 'named',
  NAMESPACE = 'namespace',
  REQUIRE = 'require',
  SIDE_EFFECT = 'side-effect'
}

export default SyntaxKind
