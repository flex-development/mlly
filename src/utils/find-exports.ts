/**
 * @file findExports
 * @module mlly/utils/findExports
 */

import { SpecifierKind, StatementKind, StatementSyntaxKind } from '#src/enums'
import type { ExportStatement } from '#src/interfaces'
import validateString from '#src/internal/validate-string'
import type { Declaration } from '#src/types'
import type { NodeError } from '@flex-development/errnode'
import {
  EXPORT_AGGREGATE_REGEX,
  EXPORT_DECLARATION_REGEX,
  EXPORT_DEFAULT_REGEX,
  EXPORT_LIST_REGEX
} from '@flex-development/export-regex'

/**
 * Finds all export statements in `code`. Ignores matches in comments.
 *
 * @see {@linkcode ExportStatement}
 * @see https://regex101.com/r/JtvRUt
 * @see https://regex101.com/r/8HpMrA
 * @see https://regex101.com/r/G7GhEt
 * @see https://regex101.com/r/KQEDdZ
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/export
 *
 * @param {string} [code=''] - Code to evaluate
 * @return {ExportStatement[]} Export statement objects
 * @throws {NodeError<TypeError>} If `code` is not a string
 */
const findExports = (code: string = ''): ExportStatement[] => {
  validateString(code, 'code')

  /**
   * `export` statement objects.
   *
   * @const {ExportStatement[]} statements
   */
  const statements: ExportStatement[] = []

  // get aggregate export statements
  for (const match of code.matchAll(EXPORT_AGGREGATE_REGEX)) {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const { exports = '', specifier = '', type = '' } = groups

    statements.push({
      code,
      declaration: null,
      end: start + code.length,
      exports: exports.startsWith('*')
        ? [exports]
        : exports
            .replace(/^{|}$/g, '')
            .split(',')
            .map(e => e.trim())
            .filter(e => e.length > 0),
      kind: StatementKind.EXPORT,
      modifiers: [],
      specifier,
      specifier_kind: SpecifierKind.STATIC,
      start,
      syntax: exports.startsWith('{')
        ? StatementSyntaxKind.NAMED
        : StatementSyntaxKind.NAMESPACE,
      type: !!type
    })
  }

  // get declaration export statements
  for (const match of code.matchAll(EXPORT_DECLARATION_REGEX)) {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const { declaration = '', exports = '', modifiers = '' } = groups

    statements.push({
      code,
      declaration: declaration as Declaration,
      end: start + code.length,
      exports: /^\w+$/.test(exports)
        ? [exports]
        : exports
            .replace(/^[[{]|[\]}]$/g, '')
            .split(',')
            .map(e => e.trim().replace(/\s*=\s*.*$/, ''))
            .filter(e => e.length > 0),
      kind: StatementKind.EXPORT,
      modifiers:
        modifiers === ''
          ? []
          : modifiers
              .split(' ')
              .map(e => e.trim())
              .filter(e => e.length > 0),
      specifier: null,
      specifier_kind: null,
      start,
      syntax: StatementSyntaxKind.DECLARATION,
      type: false
    })
  }

  // get export default statements
  for (const match of code.matchAll(EXPORT_DEFAULT_REGEX)) {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const { exports = '', kind: declaration = '', modifiers = '' } = groups

    statements.push({
      code,
      declaration: declaration ? (declaration as Declaration) : null,
      end: start + code.length,
      exports: exports === '' ? [] : [exports],
      kind: StatementKind.EXPORT,
      modifiers:
        modifiers === ''
          ? []
          : modifiers
              .split(' ')
              .map(e => e.trim())
              .filter(e => e.length > 0),
      specifier: null,
      specifier_kind: null,
      start,
      syntax: StatementSyntaxKind.DEFAULT,
      type: false
    })
  }

  // get list export statements
  for (const match of code.matchAll(EXPORT_LIST_REGEX)) {
    const { 0: code = '', index: start = 0, groups = {} } = match
    const { exports = '', type = '' } = groups

    statements.push({
      code,
      declaration: null,
      end: start + code.length,
      exports: exports
        .replace(/^{|}$/g, '')
        .split(',')
        .map(e => e.trim())
        .filter(e => e.length > 0),
      kind: StatementKind.EXPORT,
      modifiers: [],
      specifier: null,
      specifier_kind: null,
      start,
      syntax: StatementSyntaxKind.LIST,
      type: !!type
    })
  }

  return statements
}

export default findExports
