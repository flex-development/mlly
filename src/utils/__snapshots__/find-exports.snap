// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:utils/findExports > should return ExportStatement object array 1`] = `
[
  {
    "code": "export { defineBuildConfig, type BuildConfig } from \\"#src\\"",
    "declaration": null,
    "end": 117,
    "exports": [
      "defineBuildConfig",
      "type BuildConfig",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": "#src",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 59,
    "syntax": "named",
    "type": false,
  },
  {
    "code": "export type {
  JsonObject,
  LiteralUnion,
  Nullable
} from '@flex-development/tutils'",
    "declaration": null,
    "end": 206,
    "exports": [
      "JsonObject",
      "LiteralUnion",
      "Nullable",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": "@flex-development/tutils",
    "specifier_kind": "bare",
    "specifier_syntax": "static",
    "start": 118,
    "syntax": "named",
    "type": true,
  },
  {
    "code": "export {
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
} from './lib'",
    "declaration": null,
    "end": 329,
    "exports": [
      "addFive",
      "addFour",
      "addThree",
      "addTwo",
      "squareFive",
      "squareFour",
      "squareThree",
      "squareTwo",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": "./lib",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 207,
    "syntax": "named",
    "type": false,
  },
  {
    "code": "export * as constants from \\"./constants\\"",
    "declaration": null,
    "end": 370,
    "exports": [
      "* as constants",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": "./constants",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 330,
    "syntax": "namespace",
    "type": false,
  },
  {
    "code": "export type { default as Options } from \\"./options\\"",
    "declaration": null,
    "end": 457,
    "exports": [
      "default as Options",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": "./options",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 406,
    "syntax": "named",
    "type": true,
  },
  {
    "code": "export * from './utils'",
    "declaration": null,
    "end": 481,
    "exports": [
      "*",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": "./utils",
    "specifier_kind": "relative",
    "specifier_syntax": "static",
    "start": 458,
    "syntax": "namespace",
    "type": false,
  },
  {
    "code": "export interface User",
    "declaration": "interface",
    "end": 503,
    "exports": [
      "User",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 482,
    "syntax": "declaration",
    "type": false,
  },
  {
    "code": "export abstract class House",
    "declaration": "class",
    "end": 534,
    "exports": [
      "House",
    ],
    "kind": "export",
    "modifiers": [
      "abstract",
    ],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 507,
    "syntax": "declaration",
    "type": false,
  },
  {
    "code": "export const { name1, name2: bar } = o",
    "declaration": "const",
    "end": 576,
    "exports": [
      "name1",
      "name2: bar",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 538,
    "syntax": "declaration",
    "type": false,
  },
  {
    "code": "export const [ name1, name2 ] = array",
    "declaration": "const",
    "end": 615,
    "exports": [
      "name1",
      "name2",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 578,
    "syntax": "declaration",
    "type": false,
  },
  {
    "code": "export default async function foo",
    "declaration": "function",
    "end": 650,
    "exports": [
      "foo",
    ],
    "kind": "export",
    "modifiers": [
      "async",
    ],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 617,
    "syntax": "default",
    "type": false,
  },
  {
    "code": "export default foo",
    "declaration": null,
    "end": 674,
    "exports": [
      "foo",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 656,
    "syntax": "default",
    "type": false,
  },
  {
    "code": "export default",
    "declaration": null,
    "end": 689,
    "exports": [],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 675,
    "syntax": "default",
    "type": false,
  },
  {
    "code": "export {
  DEFAULTS,
  plugin as default,
  type Options
}",
    "declaration": null,
    "end": 58,
    "exports": [
      "DEFAULTS",
      "plugin as default",
      "type Options",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 0,
    "syntax": "list",
    "type": false,
  },
  {
    "code": "export type { default as Options }",
    "declaration": null,
    "end": 405,
    "exports": [
      "default as Options",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 371,
    "syntax": "list",
    "type": true,
  },
  {
    "code": "export type { Config, Result }",
    "declaration": null,
    "end": 727,
    "exports": [
      "Config",
      "Result",
    ],
    "kind": "export",
    "modifiers": [],
    "specifier": null,
    "specifier_kind": null,
    "specifier_syntax": null,
    "start": 697,
    "syntax": "list",
    "type": true,
  },
]
`;
